fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,492 @@
+}
+
+impl KeyPair for Secp256r1KeyPair {
+    type PubKey = Secp256r1PublicKey;
+    type PrivKey = Secp256r1PrivateKey;
+    type Sig = Secp256r1Signature;
+
+    fn public(&'_ self) -> &'_ Self::PubKey {
+        &self.name
+    }
+
+    fn private(self) -> Self::PrivKey {
+        Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap()
+    }
+
+    #[cfg(feature = "copy_key")]
+    fn copy(&self) -> Self {
+        Secp256r1KeyPair {
+            name: self.name.clone(),
+            secret: Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap(),
+        }
+    }
+
+    fn generate<R: AllowedRng>(rng: &mut R) -> Self {
+        let privkey = ExternalSecretKey::random(rng);
+        let pubkey = ExternalPublicKey::from(&privkey);
+
+        Secp256r1KeyPair {
+            name: Secp256r1PublicKey {
+                pubkey,
+                bytes: OnceCell::new(),
+            },
+            secret: Secp256r1PrivateKey {
+                privkey,
+                bytes: OnceCell::new(),
+            },
+        }
+    }
+}
+
+impl FromStr for Secp256r1KeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl Signer<Secp256r1Signature> for Secp256r1KeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Secp256r1Signature, signature::Error> {
+        Ok(Secp256r1Signature {
+            sig: self.secret.privkey.sign(msg),
+            bytes: OnceCell::new(),
+        })
+    }
+}
+
+impl TryFrom<Secp256r1PublicKeyBytes> for Secp256r1PublicKey {
+    type Error = signature::Error;
+
+    fn try_from(bytes: Secp256r1PublicKeyBytes) -> Result<Secp256r1PublicKey, Self::Error> {
+        Secp256r1PublicKey::from_bytes(bytes.as_ref()).map_err(|_| Self::Error::new())
+    }
+}
+
+impl From<&Secp256r1PublicKey> for Secp256r1PublicKeyBytes {
+    fn from(pk: &Secp256r1PublicKey) -> Self {
+        Secp256r1PublicKeyBytes::from_bytes(pk.as_ref()).unwrap()
+    }
+}
+
+impl From<Secp256r1PrivateKey> for Secp256r1KeyPair {
+    fn from(secret: Secp256r1PrivateKey) -> Self {
+        let name = Secp256r1PublicKey::from(&secret);
+        Secp256r1KeyPair { name, secret }
+    }
+}
+
+impl Secp256r1Signature {
+    /// Recover public key(s) from signature. Either 2 or 4 potential public keys,
+    /// which could be the corresponding public key for the signature are returned.
+    /// This is based on section 4.1.6 in https://www.secg.org/sec1-v2.pdf.
+    ///
+    /// An [FastCryptoError::GeneralError] is returned if no public keys can be recovered.
+    pub fn recover(&self, msg: &[u8]) -> Result<Vec<Secp256r1PublicKey>, FastCryptoError> {
+        // This implementation is based on recover_verifying_key_from_digest_bytes in the p256 crate,
+        // but also handles the case the the x-coordinate is larger than the group order.
+
+        let (r, s) = self.sig.split_scalars();
```

Comment:
for secp256k1, we reject signature with s in the lower half of the field. perhaps to maintain the same behavior here.

Reply:
To align the schemes, that certainly makes sense. It's what Ethereum does, right, but I don't think it's part of the standard? The wycheproof tests for p256 fails if signatures with s > n/2 signatures are rejected, so we'll have to get around that if we wan't to enforce this.